src/hooks/useMutex.js

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+
+function useMutex() {
+  const queue = React.useRef([]);
+  const lock = React.useRef(false);
+
+  const wait = React.useCallback(() => {
```

Comment:
`wait` is an operation that looks at the execution queue and sees if there are any tasks that need to be run. If there are, the lock is acquired and the task is executed.